### PR TITLE
Column Defaults Implementation Part 2

### PR DESCRIPTION
**Pull request by Hydrocharged**
Here is a comprehensive set of tests for default logic. Practically everything that was added in https://github.com/liquidata-inc/go-mysql-server/pull/174 is covered here, including some edge cases. In addition, the memory table implementation was broken/insufficient in a few ways, so I patched those up.

The biggest change besides the tests is the additional pass when projecting expressions. This is required in order for defaults that reference other columns (especially those that come after) to be able to pull the correct value. This was something I noticed only after I wrote a test that wasn't behaving as expected (compared to the MySQL client). In fact, all of the changes outside of `enginetests` were due to fixing bugs that were found from testing.

### CODE REVIEW

File: enginetest/enginetests.go [hunk 1]
```diff
@@ -1700,6 +1701,507 @@ func TestTracing(t *testing.T, harness Harness) {
...
+	)
+	_, _, err = e.Query(NewContext(harness), "INSERT INTO t9 (pk) VALUES (1), (2)")
+	require.NoError(err)
+	TestQuery(t, harness, e,
+		"SELECT * FROM t9",
+		[]sql.Row{{1, "77"}, {2, "77"}},
+	)
+
+	// DATETIME/TIMESTAMP NOW/CURRENT_TIMESTAMP literal
+	TestQuery(t, harness, e,
```
Comment (zachmu) [inline enginetest/enginetests.go:1821]: You should add a TODO to come back to these tests that you can't check values on yet. We need to be able to set the clock for tests.

File: enginetest/enginetests.go [hunk 2]
```diff
@@ -1700,6 +1701,507 @@ func TestTracing(t *testing.T, harness Harness) {
 	require.Equal(expectedSpans, spanOperations)
 }
 
+func TestColumnDefaults(t *testing.T, harness Harness) {
```
Comment (zachmu) [inline enginetest/enginetests.go:1704]: Great tests. Only comment would be to replace the comments in this function with calls to t.Run()

File: memory/table.go [hunk 3]
```diff
@@ -486,41 +504,87 @@ func (t *Table) insertValueInRows(ctx *sql.Context, idx int, colDefault *sql.Col
...
+		} else {
+			droppedCol = i
 		}
 	}
 	t.schema = newSch
-	return nil
+	return droppedCol
 }
 
 func (t *Table) ModifyColumn(ctx *sql.Context, columnName string, column *sql.Column, order *sql.ColumnOrder) error {
```
Comment (zachmu) [inline memory/table.go:538]: These things are honestly the worst, thanks for fixing them

File: sql/analyzer/resolve_columns.go [hunk 4]
```diff
@@ -481,6 +481,34 @@ func indexColumns(a *Analyzer, n sql.Node, scope *Scope) map[tableCol]indexedCol
...
 		}
 	}
+	indexSchemaForDefaults := func(column *sql.Column, order *sql.ColumnOrder, sch sql.Schema) {
+		tblSch := make(sql.Schema, len(sch))
+		copy(tblSch, sch)
+		if order == nil {
+			tblSch = append(tblSch, column)
+		} else if order.First {
+			tblSch = append(sql.Schema{column}, tblSch...)
+		} else { // must be After
```
Comment (zachmu) [inline sql/analyzer/resolve_columns.go:491]: No error case for invalid after column. Make sure there is a test for this.

Comment (zachmu) [reply sql/analyzer/resolve_columns.go:491]: (should throw an error because the after column doesn't exist)

File: enginetest/enginetests.go (hunk 1 above)
Comment (Hydrocharged) [reply enginetest/enginetests.go:1821]: Great idea!

File: enginetest/enginetests.go (hunk 2 above)
Comment (Hydrocharged) [reply enginetest/enginetests.go:1704]: I should have done this the first time 🤦

File: memory/table.go (hunk 3 above)
Comment (Hydrocharged) [reply memory/table.go:538]: There was a lot of pain involved in this file lol

File: sql/analyzer/resolve_columns.go (hunk 4 above)
Comment (Hydrocharged) [reply sql/analyzer/resolve_columns.go:491]: Ohhh fantastic catch!
